[STEP3-2] 날씨 앱 (Miro, Sunny)

longlivedrgn opened the pull request:
> 안녕하세요 제이슨!(@ehgud0670)  미로와 써니 입니다 😄 이번 스텝에서는 콜렉션 뷰를 구현할 수 있는 다양한 방법에 대해서 공부해보고, async await 또한 공부해볼 수 있는 시간이었어요! 저번에 짚어주신 Swift API GuideLine에 대해서도 다시 한번 복습해 볼 수 있었어요! 감사합니다 :)

# 요구사항 (STEP3) / 핵심 구현사항
- 오토레이아웃을 활용하여 코드로만 구현
- Collection View를 활용
    - Collection View Header의 활용
    - Collection View Custom Cell의 활용
- Configuration을 활용
- DateFormatter의 활용
- RefreshControl의 활용


# 작동 화면


![Simulator Screen Recording - iPhone 14 Plus - 2023-04-08 at 19 33 41](https://user-images.githubusercontent.com/85781941/230718284-9b51723a-a070-41de-95f4-8113b9df07c0.gif)

# 고민과 해결🤔
## Collection View 구현
- 저희는 공식문서를 통하여 다양한 콜렉션 뷰 구현 방법이 있음을 알 수 있었어요! 저희가 공부해 본 콜렉션 뷰 구현 방식은 아래와 같아요
    - **DelegateFlowLayout**
    - **CompositionalLayout**
    - **DataSource**
    - **DiffableDataSource**

### 1️⃣ **Layout**
- 위의 방식을 하나하나씩 시도해보며 콜렉션 뷰를 구현하던 중, **Configuration을** 활용하라는 구현 사항을 통하여 **DelegateFlowLayout이** 아닌 **Configuration을** 활용하는 **CompositionalLayout을** 활용하기로 했어요!
- **CompositionalLayout** 속  **UICollectionLayoutListConfiguration**을 활용하여 콜렉션 뷰 속 List를 쉽게 구현할 수 있었어요.

### 2️⃣ **DataSource**
- 제일 먼저 저희가 고려한 건 **DiffableDatasource였어요**. **DiffableDataSource는** `reloadData()`가 아닌  SnapShot을 통하여 데이터의 전 상태와 현재 상태의 변화를 통하여 콜렉션 뷰를 업데이트하는 객체에요. 이를 통해서 부드러운 UI 업데이트를 할 수 있어요.
- 그러나, 저희가 생각했을 때 현재 프로젝트에서는 다이나믹한 UI 업데이트가 없으므로 기본 **DataSource를** 활용하여 콜렉션 뷰의 UI를 업데이트하는 것이 좋을 거 같다고 판단했어요.

## ViewModel 생성과 async, await의 활용
- 아래 내용은 저번 스텝 PR에서 저희가 작성한 ViewModel을 활용하게 된 과정에 async, await에 관한 이야기를 덧붙혔어요.

### 거대해진 View Controller
- 기존의 ViewController의 경우, CoreLocation을 통하여 위치를 받아오기 위해 CLLocationManagerDelegate을 채택했어요.
- API 통신으로 받아온 데이터를 Collection View에서 보여주기 위한 로직을 담고있어야했어요.

```swift
// 기존 ViewController 코드
class WeatherViewController: UIViewController {

    private let networkModel = NetworkModel()
    private lazy var network = WeatherAPIManager(networkModel: networkModel)
    private let locationDelegate = LocationManagerDelegate()	    
    
    lazy var coreLocationManger: CLLocationManager = {	
        let manager = CLLocationManager()	
        manager.desiredAccuracy = kCLLocationAccuracyKilometer	
        manager.requestWhenInUseAuthorization()	
        return manager	
    }()	


    override func viewDidLoad() {	    
        super.viewDidLoad()	   

        coreLocationManger.delegate = locationDelegate	
        network.fetchWeatherInformation(
            of: .currentWeather, 
            in: Coordinate(longitude: 126.96368972, latitude: 37.53361968)
        )	
    }
}

```

- 그러던 중, [뷰 컨트롤러는 죄가 없다](https://www.sungdoo.dev/programming/do-not-blame-mvc-1) 라는 아티클을 보게되었어요. 
해당 아티클에서 쓰여진 말처럼, 저희의 코드는 네트워크 API콜을 하고 response를 받아 View에 띄우는 것 까지 모두 View Controller이 담당하고 있었어요.

</br>

### 2. 문제 해결 시도 : Data Controller의 생성
- 문제를 인식한 뒤, View Controller가 View에 데이터를 present해주는 로직만 담기로 결심했어요. 
View들을 관리하는 View Controller처럼, Data들을 관리하는 **Data Controller** 파일을 생성하기로 했어요. 
네트워크 API를 호출하고, 결과를 받아 View에 보여주기 위한 타입으로 변환하는 로직 등 데이터 관련 로직을 Data Controller에서 구현했어요. 

- Apple Documentation의 [Implementing Modern Collection Views](https://developer.apple.com/documentation/uikit/views_and_controls/collection_views/implementing_modern_collection_views) 예제 프로젝트를 참고했어요. 
저희는 이 프로젝트의 MountainsController 파일이 Data Controller 역할을 한다고 생각했어요. MountainsController는 Collection View에 데이터를 띄워주기 위해 Mountains 데이터 셋을 만들고 있어요. 

- 따라서, 저희는 ViewController와는 다른 Data Controller 파일을 생성하게 되었어요. 
Data Controller 파일에서는 CoreLocation의 알림을 받고, API 통신과 관련된 로직이 포함되어 있어요. 또한 Collection View에서 사용할 FiveDaysForecast 구조체와 CurrentWeather 구조체가 존재해요. 

</br>

### 3. 새로운 문제 발생 : 가벼운 ViewController, 거대한 Data Controller (MVVM..🤔)
- 리팩토링을 통해 가벼운 View Controller를 얻었지만, 그에 반해 매우 많은 로직을 담은 Data Controller를 만들게 되었어요...

```swift
final class WeatherController {
    
    struct CurrentWeather: Identifiable {
        let id = UUID()
        let image: UIImage?
        let address: String
        let temperatures: Temperature
    }

    struct FiveDaysForecast: Identifiable {
        let id = UUID()
        var image: UIImage?
        let date: String
        let temperature: Double
    }
    
    private var weatherAPIManager: WeatherAPIManager?
    private let locationManager = LocationManager()
    
    weak var currentWeatherDelegate: CurrentWeatherDelegate?

    var currentWeather: CurrentWeather?
    var forecaseWeather: [FiveDaysForecast] = []

    init(networkModel: NetworkModel = NetworkModel(session: URLSession.shared)) {
        weatherAPIManager = WeatherAPIManager(networkModel: networkModel)

        locationManager.locationDelegate = self
    }

    // CLLocation -> Coordinate 변환 함수
    func makeCoordinate(from location: CLLocation) -> Coordinate {
        // ...
    }
    
    // API 통신 후 decoded 한 CurrentWeatherDTO를 CurrentWeather 타입으로 변환하는 함수
    private func makeCurrentWeather(location: CLLocation) {
         // ...
    }

    // API 통신 후 decoded 한 FiveDaysForecastDTO를 FiveDaysForecast 타입으로 변환하는 함수
    private func makeForecastWeather(location: CLLocation) {
        // ... 
    }
}

    func makeWeatherData(location: CLLocation) {
        makeCurrentWeather(location: location)
        makeForecastWeather(location: location)
    }

}

extension WeatherController: LocationDelegate {
    func send(location: CLLocation) {
        makeWeatherData(location: location)
    }

}
```


- **그리고 아키텍쳐에 대한 고민이 익숙치 않아 다른 분들에게 의견을 구한 결과, 이러한 구조가 일종의 `MVVM`라는 것을 알게되었어요.**
    - 사소한 궁금증인데, 보통 프로젝트 내에서 ViewModel 역할을 하는 타입의 네이밍에 "ViewModel"이라는 단어를 명시적으로 사용하나요? 😮


- 그리하여 기존의 Data Controller 파일을 View Model로 변경하고, 거대해진 View Model을 다시 가볍게 나눠보기로 했어요. 

</br>

### 4. 문제 해결 시도 : SRP를 지킨 View Model
- **SRP**란, 클래스는 하나의 기능만 가지며, 클래스가 제공하는 모든 서비스는 하나의 책임을 수행하는데 집중되어야 한다는 SOLID 원칙 중 하나에요. 
- 기존 코드는 WeatherController 안에 CurrentWeather와 FiveDaysForecast가 존재하고, DTO 데이터를 각 타입으로 변환하는 두 개의 서비스를 갖고 있어요. 이러한 WeatherController는 SRP 원칙을 위배하는 코드라고 생각했어요.


**CurrentWeatherViewModel.swift**
``` swift
final class CurrentWeatherViewModel {
    
    struct CurrentWeather: Identifiable {
        let id = UUID()
        let image: UIImage?
        let address: String
        let temperatures: Temperature
    }
    
    func makeCurrentAddress(locationManager: CoreLocationManager,
                            location: CLLocation,
                            completion: @escaping (String) -> Void
    ) {
        // ...
    }
    
    func makeCurrentInformation(weatherAPIManager: WeatherAPIManager?,
                                coordinate: Coordinate,
                                location: CLLocation,
                                address: String,
                                completion: @escaping (String, CurrentWeatherDTO) -> Void
    ) {
        // ...
    }
    
    func makeCurrentImage(weatherAPIManager: WeatherAPIManager?,
                          iconString: String,
                          address: String,
                          weatherData: CurrentWeatherDTO
    ) {
             // ...
            // 추후 WeatherViewModel로 전달하는 Delegate 구현
        }
    }
}

```
- CurrentWeather 타입과 그에 관련된 서비스만을 다루는 CurrentWeatherViewModel 파일이에요.

</br>

**FiveDaysForecastWeatherViewModel.swift**
``` swift
final class FiveDaysForecastWeatherViewModel {
    
    struct FiveDaysForecast: Identifiable {
            let id = UUID()
            var image: UIImage?
            let date: String
            let temperature: Double
        }
    
    func makeForecastWeather(weatherAPIManager: WeatherAPIManager?,
                                 coordinate: Coordinate,
                                 location: CLLocation,
                                 completion: @escaping (String, Day) -> Void
        ) {
             // ...
        }
        
        func makeForecastImage(weatherAPIManager: WeatherAPIManager?,
                               icon: String,
                               eachData: Day
        ) {
                // ...
                // 추후 WeatherViewModel로 전달하는 Delegate 구현

            }
        }
}
```
- FiveDaysForecast 타입과 그에 관련된 서비스만을 다루는 FiveDaysForecastWeatherViewModel 파일로 분리했어요.

</br>

**WeatherViewModel.swift**
``` swift
final class WeatherViewModel {
    
    private let fiveDaysForecastWeatherViewModel = FiveDaysForecastWeatherViewModel()
    private let currentWeatherViewModel = CurrentWeatherViewModel()
    
    private let locationManager = CoreLocationManager()
    private let weatherAPIManager: WeatherAPIManager?
    
    var fiveDaysForecastWeather: [FiveDaysForecastWeatherViewModel.FiveDaysForecast] = []
    var currentWeather: CurrentWeatherViewModel.CurrentWeather?
    
    init(networkModel: NetworkModel = NetworkModel(session: URLSession.shared)) {
        weatherAPIManager = WeatherAPIManager(networkModel: networkModel)
        
        locationManager.locationDelegate = self
    }
    
    func makeCoordinate(from location: CLLocation) -> Coordinate {
        // ...
    }
    
    func makeWeatherData(locationManager: CoreLocationManager, weatherAPIManager: WeatherAPIManager?) {
        
        // currentWeather 만드는 메서드 호출
        currentWeatherViewModel.makeCurrentAddress(
            locationManager: locationManager,
            location: location
        ) { [weak self] address in
            
            self?.currentWeatherViewModel.makeCurrentInformation(
                weatherAPIManager: weatherAPIManager,
                coordinate: coordinate,
                location: location,
                address: address
            ) { [weak self] iconString, weatherData in
                
                self?.currentWeatherViewModel.makeCurrentImage(
                    weatherAPIManager: weatherAPIManager,
                    iconString: iconString,
                    address: address,
                    weatherData: weatherData
                )
            }
        }
        
        // fiveDaysForecast 만드는 메서드 호출       
        self.fiveDaysForecastWeatherViewModel.makeForecastWeather(
            weatherAPIManager: weatherAPIManager,
            coordinate: coordinate,
            location: location
        ) { [weak self] iconString, eachData in
            
            self?.fiveDaysForecastWeatherViewModel.makeForecastImage(
                weatherAPIManager: weatherAPIManager,
                icon: iconString,
                eachData: eachData
            )
        }
    }
}

extension WeatherViewModel: LocationDelegate {
    func didUpdateLocation() {
        makeWeatherData(
            locationManager: locationManager,
            weatherAPIManager: weatherAPIManager
        )
    }
}

```
- 또한 변환한 데이터 셋을 가지며 View Controller와 통신하는 WeatherViewModel 타입을 만들었어요.
- 이렇게 길고 길었던 리팩토링이 끝났습니다! 저희는 MVVM의 구조로(물론 처음부터 설계하고 구현한 것은 아니지만,) 코드를 구현해 보았는데 제이슨이 보시기에 MVVM 구조로 적절하게 코드를 작성한 것인지, MVVM에 대해 잘못 이해한 것은 없었는지 알려주시면 감사하겠습니다😌 


## Async, Await의 활용
- 데이터 모델을 만들기 위해서는 기본적으로 여러 번의 API 호출을 거쳐야됐어요. Current Weather의 경우, 아래의 순서를 거쳐서 객체가 생성되요.
    1. CLGeocoder().reverseGeocodeLocation 메서드를 사용해 현재의 주소를 받아와요.
    2. 경도, 위치 정보를 통하여 CurrentWeatherDTO를 생성해요.
    3. CurrentWeatherDTO의 icon 프로퍼티 값을 통하여 icon Image를 받아오는 API 통신을 해요.
    4. 최종적으로 CurrentWeather 객체를 생성해요.
- 이러한 로직을 통하여 코드를 구현하니 아래와 같은 코드가 생성되었어요.

```swift
func makeCurrentWeather(location: CLLocation) {
        
        // 1. coordinate 주소 가져오기
        let coordinate = self.makeCoordinate(from: location)
        
        // 2. address 생성하기
        locationManager.changeGeocoder(location: location) { [weak self] place in
            
            guard let locality = place?.locality, let subLocality = place?.subLocality else { return }
            
            let address = "\(locality) \(subLocality)"
            
            // 3. currentWeather 가져오기
            self?.weatherAPIManager?.fetchWeatherInformation(of: .currentWeather, in: coordinate) { [weak self] data in
                let group = DispatchGroup()
                guard let weatherData = data as? CurrentWeatherDTO else { return }
                
                guard let icon = weatherData.weather.first?.icon else { return }
                
                group.enter()
                // 4. 이미지 가져오기
                self?.weatherAPIManager?.fetchWeatherImage(icon: icon) { [weak self] weatherImage in
                    
                    self?.currentWeather = CurrentWeather(image: weatherImage, address: address, temperatures: weatherData.temperature)
                    group.leave()
                }
                group.notify(queue: .main) {
                    self?.currentWeatherDelegate?.notifyToUpdateCurrentWeather()
                }
            }
        }
    }
```
- 연속적인 completionHandler로 인해 코드의 가독성이 좋지 못하다고 생각했어요.
- 이러한 장풍 코드를 개선하기 위한 방법을 고민해봤어요.
    - DispatchGroup의 활용해 return 값으로 데이터를 전달한다.
    - async, await를 활용한다.
- 저희는 위 방법 중 async, await을 활용해보기로 하였고, 아래와 같은 코드를 완성할 수 있었어요!


### **😇 async, await를 활용하여 리팩토링한 코드는 아래와 같아요.**
### `WeatherViewModel`

```swift
// WeatherViewModel
private func execute(locationManager: CoreLocationManager,
                 location: CLLocation,
                 weatherNetworkDispatcher: WeatherNetworkDispatcher) {
        
        let coordinate = self.makeCoordinate(from: location)
        
        Task {
            let address = try await currentWeatherViewModel.fetchCurrentAddress(
                locationManager: coreLocationManager,
                location: location)
            
            let currentWeatherDTO = try await currentWeatherViewModel.fetchCurrentInformation(
                weatherNetworkDispatcher: weatherNetworkDispatcher,
                coordinate: coordinate
            )
            
            let currentWeatherImage = try await currentWeatherViewModel.fetchCurrentImage(
                weatherNetworkDispatcher: weatherNetworkDispatcher,
                currentWeatherDTO: currentWeatherDTO
            )
            
            let currentWeather = currentWeatherViewModel.makeCurrentWeather(
                image: currentWeatherImage,
                address: address,
                currentWeatherDTO: currentWeatherDTO
            )
            
            self.currentWeather = currentWeather
            
            let fiveDaysForecastWeatherDTO = try await fiveDaysForecastWeatherViewModel.fetchForecastWeather(
                weatherNetworkDispatcher: weatherNetworkDispatcher,
                coordinate: coordinate
            )
            
            let fiveDaysForecastImages = try await fiveDaysForecastWeatherViewModel.fetchForecastImages(
                weatherNetworkDispatcher: weatherNetworkDispatcher,
                fiveDaysForecastDTO: fiveDaysForecastWeatherDTO
            )
            
            let fiveDaysForecasts = fiveDaysForecastWeatherViewModel.makeFiveDaysForecast(
                images: fiveDaysForecastImages,
                fiveDaysForecastDTO: fiveDaysForecastWeatherDTO
            )
            
            self.fiveDaysForecastWeather = fiveDaysForecasts
            
            DispatchQueue.main.async {
                self.delegate?.weatherViewModelDidFinishSetUp(self)
            }
        }
    }
```

### `CurrentWeatherViewModel`
```swift
func fetchCurrentAddress(locationManager: CoreLocationManager,
                             location: CLLocation) async throws -> String {
        
        let location = try await locationManager.changeGeocoder(location: location)
        
        guard let locality = location?.locality, let subLocality = location?.subLocality else {
            throw NetworkError.failedTypeCasting
        }
        
        let address = "\(locality) \(subLocality)"
        
        return address
    }
    
    func fetchCurrentInformation(weatherNetworkDispatcher: WeatherNetworkDispatcher,
                                 coordinate: Coordinate) async throws -> CurrentWeatherDTO {
        
        let decodedData = try await weatherNetworkDispatcher.requestWeatherInformation(of: .currentWeather, in: coordinate)
        
        guard let currentWeatherDTO = decodedData as? CurrentWeatherDTO else {
            throw NetworkError.failedTypeCasting
        }
        
        return currentWeatherDTO
    }
    
    func fetchCurrentImage(weatherNetworkDispatcher: WeatherNetworkDispatcher,
                           currentWeatherDTO: CurrentWeatherDTO) async throws -> UIImage {
        
        guard let iconString = currentWeatherDTO.weather.first?.icon else {
            throw NetworkError.failedTypeCasting
        }
        let image = try await weatherNetworkDispatcher.requestWeatherImage(icon: iconString)
        guard let image = image else {
            throw NetworkError.failedTypeCasting
        }
        
        return image
    }
```

### `FiveDaysForecastWeatherViewModel`
```swift
func fetchForecastWeather(weatherNetworkDispatcher: WeatherNetworkDispatcher,
                              coordinate: Coordinate) async throws -> FiveDaysForecastDTO {
        
        let decodedData = try await weatherNetworkDispatcher.requestWeatherInformation(of: .fiveDaysForecast, in: coordinate)
        
        guard let fiveDaysForecastDTO = decodedData as? FiveDaysForecastDTO else {
            throw NetworkError.failedTypeCasting
        }
        
        return fiveDaysForecastDTO
    }
    func fetchForecastImages(weatherNetworkDispatcher: WeatherNetworkDispatcher,
                             fiveDaysForecastDTO: FiveDaysForecastDTO) async throws -> [UIImage] {
        
        var images: [UIImage] = []
        
        for day in fiveDaysForecastDTO.list {
            guard let iconString = day.weather.first?.icon else {
                throw NetworkError.failedTypeCasting
            }
            
            let image = try await weatherNetworkDispatcher.requestWeatherImage(icon: iconString)
            
            guard let image = image else {
                throw NetworkError.failedTypeCasting
            }
            
            images.append(image)
        }
        
        return images
    }
```

- 이와 같이 async, await을 통하여 기존의 completion handler을 통한 수 많은 콜백 지옥에서 벗어날 수 있었어요!

longlivedrgn committed:
feat: collectionview 생성 및 layout list configuration 설정하기

SunnnySong committed:
feat: CurrentWeatherCell 생성

longlivedrgn committed:
feat: FiveDaysForecastCell 생성하기

SunnnySong committed:
feat: UICollectionView, UICollectionReusableView extension 추가 및 collectionView에 cell register

longlivedrgn committed:
feat: String extension 생성 및 Datasource 구현

- dateformatter를 string extension으로 구현
- datasource-cellForItemAt 구현하기

SunnnySong committed:
feat: collectionView에 headerview 등록

longlivedrgn committed:
feat: UICollectionView extension 추가

- dequeue 메소드 추가

longlivedrgn committed:
feat: CollectionView backgroundView 설정하기

longlivedrgn committed:
feat: 불필요한 identifiable 프로토콜 삭제

longlivedrgn committed:
feat: Delegate 설정 및 글씨 색 변경

longlivedrgn committed:
feat: WeatherViewModelDelegate 설정하기

longlivedrgn committed:
refactor: async, await을 활용하여 currentWeather Data 가져오는 코드 리팩토링

longlivedrgn committed:
feat: Network Error에 새로운 케이스 추가

longlivedrgn committed:
refactor: WeatherNetworkDispatcher, NetworkSession 메소드를 async, await로 리팩토링

longlivedrgn committed:
refactor: async, await을 활용하여 FiveDaysForecastWeatherViewModel 생성하게 리팩토링

longlivedrgn committed:
fix: FiveDaysForecastCell에서 height constraint error 해결

longlivedrgn committed:
feat: 온도 text가 솟수점 첫째자리까지 나오게 변경

longlivedrgn committed:
feat: Refresh control 추가하기

longlivedrgn committed:
chore: 들여쓰기 및 띄어쓰기 수정

longlivedrgn committed:
chore: 불필요한 delegate 코드 제거

SunnnySong committed:
Feat: String extension: ° 추가

SunnnySong committed:
refactor: Double extension: 소수점 1자리수까지 변경하는 함수 추가

ehgud0670 commented on WeatherForecast/WeatherForecast/Model/Delegates/WeatherViewModelDelegate.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+protocol WeatherViewModelDelegate: AnyObject {
```
ViewModel, Model의 경우 Delegate 패턴을 사용하는게 이상하지 않을까요~?
제가 알기론 View나 ViewController의 경우엔 Delegate 패턴을 사용하고, ViewModel, Model 의 경우 Observer 패턴(ex. `NotificationCenter` Or `RxSwift`, `Combine`) 을 사용하는 걸로 알고 있습니다.

longlivedrgn replied:
기존의 delegate pattern이 아닌 observer pattern인 `NotificationCenter`를 활용하여 리팩토링 해봤습니다!

ehgud0670 commented on WeatherForecast/WeatherForecast/Views/Cells/CurrentWeatherCell.swift:
```diff
@@ -0,0 +1,53 @@
+       
+       override func updateConfiguration(using state: UICellConfigurationState) {
+           super.updateConfiguration(using: state)
+           
```
여기 정렬이 이상하네요.

longlivedrgn replied:
들여쓰기 수정했습니다!☺️

ehgud0670 commented on WeatherForecast/WeatherForecast/Views/Cells/CurrentWeatherCell.swift:
```diff
@@ -0,0 +1,53 @@
+
+final class CurrentWeatherCell: UICollectionViewListCell {
+    
+    var currentWeather: CurrentWeatherViewModel.CurrentWeather?
```
CurrentWeatherCell이 currentWeather 모델 값을 아예 갖는건 의존성이 너무 높아 보여요. currentWeather를 프로퍼티로 갖는게 아니라 메소드의 파라미터로 currentWeather을 가져와서 UI를 업데이트 해보세요.

첨언하자면, 
* model 값 필요없이 UI setup 하는 메소드 
* model 값 필요한 UI update 하는 메소드 
두개로 나누어 리팩토링 해보세요.

ehgud0670 replied:
말씀주신대로 해당 셀은 UpdateConfiguration이라는 메소드를 써서 모델값을 들고 있을수 밖에 없네요. 알겠습니다 :)

ehgud0670 commented on WeatherForecast/WeatherForecast/Views/Cells/CurrentWeatherCell.swift:
```diff
@@ -0,0 +1,53 @@
+                 let currentTemperature = currentWeather?.temperatures.averageTemperature
+           else { return }
+
+           let addressAndTemperatureText: String = """
+           \(address)
+           최저 \(minimumTemperature.changeWeatherFormat().degree) 최소 \(maximumTemperature.changeWeatherFormat().degree)
+           """
```
이 text 값도 CurrentWeather 모델의 연산프로퍼티나 프로퍼티로 빼보는건 어떨까요~?

longlivedrgn replied:
CurrentWeather 모델 안에 CustomStringConvertible을 채택하여 description 연산 프로퍼티를 활용했어요! 그랬더니 cell 안에서 해주는 일이 훨씬 줄어들어 좋은 거 같습니다!

ehgud0670 commented on WeatherForecast/WeatherForecast/Views/Cells/FiveDaysForecastCell.swift:
```diff
@@ -0,0 +1,62 @@
+
+class FiveDaysForecastCell: UICollectionViewCell {
+    
+    var weatherIconImage: UIImageView = {
+        let imageView = UIImageView()
+        imageView.contentMode = .scaleAspectFill
+        imageView.backgroundColor = .clear
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
+    
+    var dateLabel: UILabel = {
+        let label = UILabel()
+        label.font = .systemFont(ofSize: 16)
+        label.textColor = .white
+        label.translatesAutoresizingMaskIntoConstraints = false
+        return label
+    }()
+    
+    var temperatureLabel: UILabel = {
+        let label = UILabel()
+        label.font = .systemFont(ofSize: 16)
+        label.textColor = .white
+        label.translatesAutoresizingMaskIntoConstraints = false
+        return label
+    }()
```
여기 프로퍼티들 모두 let 으로 해도 되지 않아요?

ehgud0670 replied:
private 으로 모두 감추고 메소드로 UI 업데이트 하도록 리팩토링 해보시는건 어떠실까요~?

SunnnySong replied:
1. 모든 프로퍼티 private let으로 처리했습니다!
2. dateLabel과 temperatureLabel에 중복되는 코드가 있어 extension으로 convenience init을 추가해주었습니다!
3. 외부에서 받아온 fiveDaysForecast 데이터를 각 프로퍼티와 연결하는 `configure(with:)` 함수를 생성했습니다.
4. layout을 잡는 함수를 `setupConstraints()`로, addSubview하는 함수를 `setup()`으로 분리하였습니다!

ehgud0670 commented on WeatherForecast/WeatherForecast/Views/Cells/FiveDaysForecastCell.swift:
```diff
@@ -0,0 +1,62 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    func setUp() {
```
```suggestion
    private func setUp() {
```

SunnnySong replied:
수정하였습니다~~

ehgud0670 commented on WeatherForecast/WeatherForecast/Views/WeatherViewController.swift:
```diff
@@ -10,9 +10,118 @@ import CoreLocation
 class WeatherViewController: UIViewController {
     
     private var weatherViewModel = WeatherViewModel()
```
불필요하게 var 를 쓴 곳이 많네요.

longlivedrgn replied:
변경되지 않는 변수는 let으로 바꿔주었습니다! 감사합니다 :)

ehgud0670 commented on WeatherForecast/WeatherForecast/Views/WeatherViewController.swift:
```diff
@@ -10,9 +10,118 @@ import CoreLocation
+        weatherCollectionView.backgroundView = backgroundImageView
+        configureRefreshControl(in: weatherCollectionView)
+        view.addSubview(weatherCollectionView)
+    }
```
메소드 명이 모호한것 같아요 계층을 configure 한다고 하는데 뷰들의 attribute 들을 setup 하는 코드도 섞여있네요. 
차라리 configurUI로 작명하는 건 어떨까요? 아니면 configurAttributes, configureLayouts으로 메소드 나눠도 되고요.

longlivedrgn replied:
- collectionView의 BackgroundView 설정 및 refreshControl 생성하는 함수를 configureAttributes 함수로 명명
- view에 addSubView하는 함수를 setUp함수로 명명 

위와 같이 함수를 분리하여 명명했습니다!

ehgud0670 reviewed (changes requested):
고생하셨습니다. 
피드백 확인하여 반영 부탁드려요 🙏

ps. async, await 으로 잘 바꾸셨네요 👏👏👏

longlivedrgn committed:
fix: 들여쓰기 수정

SunnnySong committed:
refactor: FiveDaysForecastCell 리뷰 반영

- 프로퍼티 let 변경
- cell 프로퍼티의 값을 외부에서 던저주면 FiveDaysForecastCell 내부에서 연결
- UILabel에서 extension으로 init 추가
- autoLayout 잡는 코드를 setupConstraints로 함수 분리

SunnnySong committed:
Merge branch 'main' of https://github.com/longlivedrgn/ios-weather-forecast

longlivedrgn committed:
refactor: 불필요한 var let으로 변경

longlivedrgn committed:
feat: weatherViewController 속 configureHierarchy 함수를 세부 함수로 분리

longlivedrgn committed:
refactor: 기존의 delegate pattern을 observer pattern으로 변경

SunnnySong committed:
refactor: CurrentWeatherCell 코드 분리

SunnnySong committed:
Merge branch 'main' of https://github.com/longlivedrgn/ios-weather-forecast

SunnnySong committed:
refactor: CurrentWeather에 description 추가

ehgud0670 commented on WeatherForecast/WeatherForecast/ViewModels/WeatherViewModel.swift:
```diff
@@ -81,7 +81,7 @@ final class WeatherViewModel {
             
             DispatchQueue.main.async {
-                self.delegate?.weatherViewModelDidFinishSetUp(self)
+                NotificationCenter.default.post(name: Notification.Name.modelDidFinishSetUp, object: nil)
```
노티를 보내는 건 백그라운드 스레드에서 하고, UI 를 업데이트 하는 코드만 메인스레드에서 돌아가게 하는게 좀 더 효율적이지 않을지 고민해보세용 ㅎㅎ

longlivedrgn replied:
넵 감사합니다! Notification은 백그라운드 스레드에서 해결을 하고, reloadData()만 메인 스레드에서 돌아가게 코드를 변경했습니다!

ehgud0670 reviewed (approved):
스레드 관련해서만 추가 코멘트 드렸어요. 
그럼 머지 하도록 하겠습니다.
고생 많으셨습니다~!! 👍👏💯
